main/lsp/LSPTask.cc

```diff
@@ -128,4 +243,41 @@ void LSPTask::addLocIfExists(const core::GlobalState &gs, vector<unique_ptr<Loca
     }
 }
 
+LSPQueuePreemptionTask::LSPQueuePreemptionTask(const LSPConfiguration &config, absl::Notification &finished,
+                                               absl::Mutex &taskQueueMutex, TaskQueueState &taskQueue,
+                                               LSPIndexer &indexer)
+    : LSPTask(config, LSPMethod::SorbetError), finished(finished), taskQueueMutex(taskQueueMutex), taskQueue(taskQueue),
+      indexer(indexer) {}
+
+void LSPQueuePreemptionTask::run(LSPTypecheckerDelegate &tc) {
+    for (;;) {
+        unique_ptr<LSPTask> task;
+        {
+            absl::MutexLock lck(&taskQueueMutex);
+            if (taskQueue.pendingTasks.empty() || !taskQueue.pendingTasks.front()->canPreempt(indexer)) {
+                break;
+            }
+            task = move(taskQueue.pendingTasks.front());
+            taskQueue.pendingTasks.pop_front();
+
+            // Index while holding lock to prevent races with processing thread.
+            task->index(indexer);
```

Comment:
This is perhaps the main tricky bit of this method.

The processing thread uses LSPIndex to determine if the head of the task queue can preempt (aka take the fast path); if it cannot, it cancels the scheduled preemption task.

The preemption task _mutates_ LSPIndex to commit edits before it typechecks them.

By doing both operations while holding the task queue mutex, we linearize them and avoid races.